node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1552 @@
+				Some(occupied.para_id)
+			} else {
+				None
+			}
+		})
+		.collect();
+	Ok(occupied_para_ids)
+}
+
+/// Modify the reputation of a peer based on its behaviour.
+async fn modify_reputation<Context>(ctx: &mut Context, peer: PeerId, rep: Rep) -> Result<()>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	trace!(
+		target: TARGET,
+		"Reputation change of {:?} for peer {:?}",
+		rep,
+		peer
+	);
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::ReportPeer(peer, rep),
+	))
+	.await
+	.map_err::<Error, _>(Into::into)
+}
+
+/// Query the proof of validity for a particular candidate hash.
+async fn query_available_data<Context>(
+	ctx: &mut Context,
+	candidate_hash: H256,
+) -> Result<Option<AvailableData>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(AllMessages::AvailabilityStore(
+			AvailabilityStoreMessage::QueryAvailableData(candidate_hash, tx),
+		))
+		.await
+		.map_err::<Error, _>(Into::into)?;
+	rx.await.map_err::<Error, _>(Into::into)
+}
+
+async fn store_chunk<Context>(
+	ctx: &mut Context,
+	candidate_hash: H256,
+	validator_index: ValidatorIndex,
+	erasure_chunk: ErasureChunk,
+) -> Result<std::result::Result<(), ()>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(AllMessages::AvailabilityStore(
+		AvailabilityStoreMessage::StoreChunk(candidate_hash, validator_index, erasure_chunk, tx),
+	))
+	.await
+	.map_err::<Error, _>(Into::into)?;
+	rx.await.map_err::<Error, _>(Into::into)
+}
+
+/// Request the head data for a particular para.
+async fn query_pending_availability<Context>(
+	ctx: &mut Context,
+	relay_parent: H256,
+	para: ParaId,
+) -> Result<Option<CommittedCandidateReceipt>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(AllMessages::RuntimeApi(RuntimeApiMessage::Request(
+		relay_parent,
+		RuntimeApiRequest::CandidatePendingAvailability(para, tx),
+	)))
+	.await
+	.map_err::<Error, _>(Into::into)?;
+	rx.await
+		.map_err::<Error, _>(Into::into)?
+		.map_err::<Error, _>(Into::into)
+}
+
+/// Query the validator set.
+async fn query_validators<Context>(
+	ctx: &mut Context,
+	relay_parent: H256,
+) -> Result<Vec<ValidatorId>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	let query_validators = AllMessages::RuntimeApi(RuntimeApiMessage::Request(
+		relay_parent.clone(),
+		RuntimeApiRequest::Validators(tx),
+	));
+
+	ctx.send_message(query_validators)
+		.await
+		.map_err::<Error, _>(Into::into)?;
+	rx.await
+		.map_err::<Error, _>(Into::into)?
+		.map_err::<Error, _>(Into::into)
+}
+
+
+/// Query the hash of the `K` ancestors
+#[cfg(feature = "std")]
+async fn query_k_ancestors<Context>(
```

Comment:
Can we make this function actually do "up to k ancestors that have the same session as parent"?

Reply:
This would use the `session_index_for_child`, so actually it's "up to k ancestors that have the same session as the parent's child". The implication is that there might just be no candidates to distribute for a given relay-parent.

This mirrors the behavior of the runtime. If you have a block B which is the last block of the session, anything pending availability will be discarded in the `on_initialize` logic of its child. So there is no use distributing anything pending availability in such a block anyway.

Reply:
However I notice that the treatment of the relay-parent itself is separate from the ancestors in `query_live_candidates`, so it might make sense to re-structure that code somewhat to perform these two steps:
  1. get a set of relay-chain block hashes, which may be empty, that contains all of the relay-parent and its K ancestors who have the same session as the relay-parent's child. This implies making a state query at `relay_parent - (K+1)`
  2. Get live candidates for each of those, querying the cache as necessary.

Reply:
> However I notice that the treatment of the relay-parent itself is separate from the ancestors in `query_live_candidates`, so it might make sense to re-structure that code somewhat to perform these two steps:
> 
>     1. get a set of relay-chain block hashes, which may be empty, that contains all of the relay-parent and its K ancestors who have the same session as the relay-parent's child. This implies making a state query at `relay_parent - (K+1)`

Don't we have to do a binary search/linear traversal over `{ relay_parant, ancestor_1, ancestor_2, .. ancestor_N } ` to know which one actually is the last one in the same session? 

One simple probe would not tell us which ancestors have the same session index.

Reply:
linear traversal would make most sense given that `K` is so small